src/fastpath/fastpath.c

```diff
@@ -902,3 +902,240 @@ void NORETURN fastpath_vm_fault(vm_fault_type_t type)
     fastpath_restore(badge, msgInfo, NODE_STATE(ksCurThread));
 }
 #endif
+
+#ifdef CONFIG_IRQ_FASTPATH
+#ifdef CONFIG_ARCH_ARM
+static inline
+FORCE_INLINE
+#endif
+void NORETURN fastpath_irq()
+{
+
+    sched_context_t *sc = NULL;
+    bool_t schedulable = false;
+    bool_t idle = false;
+    tcb_t *dest = NULL;
+    cap_t newVTable;
+    vspace_root_t *cap_pd;
+    pde_t stored_hw_asid;
+
+    /* The IRQ should be passed as a parameter to this fn. At the moment, if the
+    slowpath is called, calls getActiveIRQ again in handleInterruptEntry.
+    Syscall API change to handleInterruptEntry will be done in another commit. */
+    irq_t irq = getActiveIRQ();
+
+    /* check the irq is valid */
+    if (unlikely(IRQT_TO_IRQ(irq) == IRQT_TO_IRQ(irqInvalid))) {
+#ifdef CONFIG_IRQ_REPORTING
+        printf("Spurious interrupt\n");
+#endif
+        handleSpuriousIRQ();
+        restore_user_context();
+        UNREACHABLE();
+    }
+
+    if (unlikely(IRQT_TO_IRQ(irq) > maxIRQ)) {
+        maskInterrupt(true, irq);
+        ackInterrupt(irq);
+        restore_user_context();
+        UNREACHABLE();
+    }
+
+    /* only fastpath signalling IRQs*/
+    if (unlikely(intStateIRQTable[IRQT_TO_IDX(irq)] != IRQSignal)) {
```

Comment:
Do you reckon we should include the MCS timer case in the fastpath? It seems fairly important for that one to be performant I reckon. Maybe it won't show up on benchmarks, but I feel like if that is slower it might weaken the realtime-ness of the kernel in some way? Might be worth including for now and removing if people think it is unnecessary or at least having a comment or something to indicate this is something that might be worth considering in the real PR.